Author: Tian Shilin <tianshilin@uniontech.com>
Date:   Fri, 24 Apr 2026 10:11:37 +0300
Subject: Fix stack overflow caused by recursive calls of destructors
Upstream: https://codereview.qt-project.org/c/qt/qtsvg/+/736142
Bug: QTBUG-145916

Index: qt6-svg/src/svg/qsvgstructure.cpp
===================================================================
--- qt6-svg.orig/src/svg/qsvgstructure.cpp
+++ qt6-svg/src/svg/qsvgstructure.cpp
@@ -604,6 +604,44 @@ QSvgNode* QSvgStructureNode::previousSib
     return prev;
 }
 
+void QSvgStructureNode::releaseDescendants()
+{
+    // This function will release the descendants of a QSvgStructureNode from bottom to top.
+    // Destructors are never called recursively in this case and stack overflow will not
+    // happen in deeply nested trees.
+    // This function does not allocate any memory at the cost of sacrificing some performance to
+    // make it safe to be called from a destructor.
+    while (!m_renderers.empty()) {
+        auto nodes = &m_renderers;
+        bool isSubtree = true;
+        while (isSubtree) {
+            switch (nodes->front()->type()) {
+            case QSvgNode::Doc:
+            case QSvgNode::Defs:
+            case QSvgNode::Group:
+            case QSvgNode::Mask:
+            case QSvgNode::Pattern:
+            case QSvgNode::Symbol:
+            case QSvgNode::Switch:
+            case QSvgNode::Filter:
+            {
+                QSvgStructureNode *subtree = static_cast<QSvgStructureNode *>(nodes->first());
+                isSubtree = !subtree->m_renderers.empty();
+                if (isSubtree)
+                    nodes = &subtree->m_renderers;
+            }
+                break;
+            default:
+                isSubtree = false;
+                break;
+            }
+        }
+        QSvgNode *node = nodes->first();
+        delete node;
+        nodes->pop_front();
+    }
+}
+
 QSvgMask::QSvgMask(QSvgNode *parent, QSvgRectF bounds,
                    QtSvg::UnitTypes contentUnits)
     : QSvgStructureNode(parent)
Index: qt6-svg/src/svg/qsvgstructure_p.h
===================================================================
--- qt6-svg.orig/src/svg/qsvgstructure_p.h
+++ qt6-svg/src/svg/qsvgstructure_p.h
@@ -37,6 +37,10 @@ public:
     QRectF internalBounds(QPainter *p, QSvgExtraStates &states) const override;
     QSvgNode *previousSiblingNode(QSvgNode *n) const;
     QList<QSvgNode*> renderers() const { return m_renderers; }
+
+protected:
+    void releaseDescendants();
+
 protected:
     QList<QSvgNode*>          m_renderers;
     QHash<QString, QSvgNode*> m_scope;
Index: qt6-svg/src/svg/qsvgtinydocument.cpp
===================================================================
--- qt6-svg.orig/src/svg/qsvgtinydocument.cpp
+++ qt6-svg/src/svg/qsvgtinydocument.cpp
@@ -37,6 +37,10 @@ QSvgTinyDocument::QSvgTinyDocument(QtSvg
 
 QSvgTinyDocument::~QSvgTinyDocument()
 {
+    // Only do that when AssumeTrustedSource is set to false. Otherwise, all nodes
+    // will be deleted by recursive calls of destructors.
+    if (!m_states.trustedSource)
+        releaseDescendants();
 }
 
 static bool hasSvgHeader(const QByteArray &buf)
