Author: Tian Shilin <tianshilin@uniontech.com>
Date:   Wed, 22 Apr 2026 13:46:59 +0300
Subject: Fix rendering causing stack overflow
Upstream: https://codereview.qt-project.org/c/qt/qtsvg/+/736141
Bug: QTBUG-145916

Index: qt6-svg/src/svg/qsvgnode.cpp
===================================================================
--- qt6-svg.orig/src/svg/qsvgnode.cpp
+++ qt6-svg/src/svg/qsvgnode.cpp
@@ -42,6 +42,16 @@ void QSvgNode::draw(QPainter *p, QSvgExt
 #endif
 
     if (shouldDrawNode(p, states)) {
+        quint8 remainingDepth = states.trustedSource ? states.remainingNestedNodes
+                                                     : states.remainingNestedNodes - 1;
+        QScopedValueRollback<quint8> nestedNodesGuard(states.remainingNestedNodes, remainingDepth);
+        if (states.remainingNestedNodes == 0) {
+            qCWarning(lcSvgDraw) << "Too many nested nodes at" << qPrintable(typeName())
+                                 << "exceeding max nested limit of" << QtSvg::renderingMaxNestedNodes << "."
+                                 << "Enable AssumeTrustedSource in QSvgHandler or set QT_SVG_DEFAULT_OPTIONS=2 to disable this check.";
+            return;
+        }
+
         applyStyle(p, states);
         QSvgNode *maskNode = this->hasMask() ? document()->namedNode(this->maskId()) : nullptr;
         QSvgFilterContainer *filterNode = this->hasFilter() ? static_cast<QSvgFilterContainer*>(document()->namedNode(this->filterId()))
@@ -625,7 +635,7 @@ bool QSvgNode::shouldDrawNode(QPainter *
     if (m_displayMode == DisplayMode::NoneMode)
         return false;
 
-    if (document() && document()->options().testFlag(QtSvg::AssumeTrustedSource))
+    if (document() && states.trustedSource)
         return true;
 
     QRectF brect = internalFastBounds(p, states);
Index: qt6-svg/src/svg/qsvgstyle_p.h
===================================================================
--- qt6-svg.orig/src/svg/qsvgstyle_p.h
+++ qt6-svg/src/svg/qsvgstyle_p.h
@@ -115,6 +115,8 @@ struct Q_SVG_EXPORT QSvgExtraStates
     qreal strokeDashOffset;
     int nestedUseLevel = 0;
     int nestedUseCount = 0;
+    bool trustedSource = false;
+    quint8 remainingNestedNodes = QtSvg::renderingMaxNestedNodes;
     bool vectorEffect; // true if pen is cosmetic
     qint8 imageRendering; // QSvgQualityStyle::ImageRendering
     bool inUse = false; // true if currently in QSvgUseNode
Index: qt6-svg/src/svg/qtsvgglobal_p.h
===================================================================
--- qt6-svg.orig/src/svg/qtsvgglobal_p.h
+++ qt6-svg/src/svg/qtsvgglobal_p.h
@@ -28,6 +28,7 @@ enum class UnitTypes : quint32 {
     userSpaceOnUse
 };
 
+constexpr quint8 renderingMaxNestedNodes = 32;
 }
 
 QT_END_NAMESPACE
Index: qt6-svg/src/svg/qsvgtinydocument.cpp
===================================================================
--- qt6-svg.orig/src/svg/qsvgtinydocument.cpp
+++ qt6-svg/src/svg/qsvgtinydocument.cpp
@@ -33,6 +33,7 @@ QSvgTinyDocument::QSvgTinyDocument(QtSvg
     , m_fps(30)
     , m_options(options)
 {
+    m_states.trustedSource = m_options.testFlag(QtSvg::AssumeTrustedSource);
 }
 
 QSvgTinyDocument::~QSvgTinyDocument()
