Author: Tian Shilin <tianshilin@uniontech.com>
Date:   Wed, 22 Apr 2026 08:57:44 +0300
Subject: Fix resolving paint servers going through stack overflow
Upstream: https://codereview.qt-project.org/c/qt/qtsvg/+/730501
Bug: QTBUG-145916

Index: qt6-svg/src/svg/qsvghandler.cpp
===================================================================
--- qt6-svg.orig/src/svg/qsvghandler.cpp
+++ qt6-svg/src/svg/qsvghandler.cpp
@@ -1036,7 +1036,7 @@ static void parseBrush(QSvgNode *node,
                 } else {
                     QString id = idFromUrl(value);
                     prop->setPaintStyleId(id);
-                    prop->setPaintStyleResolved(false);
+                    handler->pushUnresolvedStyle(prop);
                 }
             } else if (attributes.fill != QLatin1String("none")) {
                 QColor color;
@@ -1206,7 +1206,7 @@ static void parsePen(QSvgNode *node,
                     } else {
                         QString id = idFromUrl(value);
                         prop->setPaintStyleId(id);
-                        prop->setPaintStyleResolved(false);
+                        handler->pushUnresolvedStyle(prop);
                     }
             } else if (attributes.stroke != QLatin1String("none")) {
                 QColor color;
@@ -4668,7 +4668,7 @@ void QSvgHandler::parse()
             break;
         }
     }
-    resolvePaintServers(m_doc);
+    resolvePaintServers();
     resolveNodes();
     if (detectCycles(m_doc)) {
         qCWarning(lcSvgHandler, "Cycles detected in SVG, document discarded.");
@@ -4902,33 +4902,23 @@ bool QSvgHandler::endElement(const QStri
     return ((localName == QLatin1String("svg")) && (node != Doc));
 }
 
-void QSvgHandler::resolvePaintServers(QSvgNode *node, int nestedDepth)
+void QSvgHandler::resolvePaintServers()
 {
-    if (!node || (node->type() != QSvgNode::Doc && node->type() != QSvgNode::Group
-        && node->type() != QSvgNode::Defs && node->type() != QSvgNode::Switch)) {
-        return;
-    }
-
-    QSvgStructureNode *structureNode = static_cast<QSvgStructureNode *>(node);
-
-    const QList<QSvgNode *> ren = structureNode->renderers();
-    for (auto it = ren.begin(); it != ren.end(); ++it) {
-        QSvgFillStyle *fill = static_cast<QSvgFillStyle *>((*it)->styleProperty(QSvgStyleProperty::FILL));
-        if (fill && !fill->isPaintStyleResolved()) {
+    for (QSvgStyleProperty *prop : std::as_const(m_unresolvedStyles)) {
+        if (prop->type() == QSvgStyleProperty::FILL) {
+            QSvgFillStyle *fill = static_cast<QSvgFillStyle *>(prop);
             QString id = fill->paintStyleId();
-            QSvgPaintStyleProperty *style = structureNode->styleProperty(id);
+            QSvgPaintStyleProperty *style = m_doc->namedStyle(id);
             if (style) {
                 fill->setFillStyle(style);
             } else {
                 qCWarning(lcSvgHandler, "%s", msgCouldNotResolveProperty(id, xml).constData());
                 fill->setBrush(Qt::NoBrush);
             }
-        }
-
-        QSvgStrokeStyle *stroke = static_cast<QSvgStrokeStyle *>((*it)->styleProperty(QSvgStyleProperty::STROKE));
-        if (stroke && !stroke->isPaintStyleResolved()) {
+        } else if (prop->type() == QSvgStyleProperty::STROKE) {
+            QSvgStrokeStyle *stroke = static_cast<QSvgStrokeStyle *>(prop);
             QString id = stroke->paintStyleId();
-            QSvgPaintStyleProperty *style = structureNode->styleProperty(id);
+            QSvgPaintStyleProperty *style = m_doc->namedStyle(id);
             if (style) {
                 stroke->setStyle(style);
             } else {
@@ -4936,10 +4926,9 @@ void QSvgHandler::resolvePaintServers(QS
                 stroke->setStroke(Qt::NoBrush);
             }
         }
-
-        if (nestedDepth < 2048)
-            resolvePaintServers(*it, nestedDepth + 1);
     }
+
+    m_unresolvedStyles.clear();
 }
 
 void QSvgHandler::resolveNodes()
@@ -5055,6 +5044,11 @@ QColor QSvgHandler::currentColor() const
         return QColor(0, 0, 0);
 }
 
+void QSvgHandler::pushUnresolvedStyle(QSvgStyleProperty *prop)
+{
+    m_unresolvedStyles.append(prop);
+}
+
 #ifndef QT_NO_CSSPARSER
 
 void QSvgHandler::setInStyle(bool b)
Index: qt6-svg/src/svg/qsvghandler_p.h
===================================================================
--- qt6-svg.orig/src/svg/qsvghandler_p.h
+++ qt6-svg/src/svg/qsvghandler_p.h
@@ -82,6 +82,8 @@ public:
     void popColor();
     QColor currentColor() const;
 
+    void pushUnresolvedStyle(QSvgStyleProperty *prop);
+
 #ifndef QT_NO_CSSPARSER
     void setInStyle(bool b);
     bool inStyle() const;
@@ -118,6 +120,7 @@ private:
     // - <use> nodes which haven't been resolved yet.
     // - <filter> nodes to be checked for unsupported filter primitives.
     QList<QSvgNode *> m_toBeResolved;
+    QList<QSvgStyleProperty *> m_unresolvedStyles;
 
     enum CurrentNode
     {
@@ -150,7 +153,7 @@ private:
     QCss::Parser m_cssParser;
 #endif
     void parse();
-    void resolvePaintServers(QSvgNode *node, int nestedDepth = 0);
+    void resolvePaintServers();
     void resolveNodes();
 
     QPen m_defaultPen;
Index: qt6-svg/src/svg/qsvgstyle.cpp
===================================================================
--- qt6-svg.orig/src/svg/qsvgstyle.cpp
+++ qt6-svg/src/svg/qsvgstyle.cpp
@@ -94,7 +94,6 @@ QSvgFillStyle::QSvgFillStyle()
     , m_oldFillRule(Qt::WindingFill)
     , m_fillOpacity(1.0)
     , m_oldFillOpacity(0)
-    , m_paintStyleResolved(1)
     , m_fillRuleSet(0)
     , m_fillOpacitySet(0)
     , m_fillSet(0)
@@ -249,7 +248,6 @@ QSvgStrokeStyle::QSvgStrokeStyle()
     , m_strokeDashOffset(0)
     , m_oldStrokeDashOffset(0)
     , m_style(0)
-    , m_paintStyleResolved(1)
     , m_vectorEffect(0)
     , m_oldVectorEffect(0)
     , m_strokeSet(0)
Index: qt6-svg/src/svg/qsvgstyle_p.h
===================================================================
--- qt6-svg.orig/src/svg/qsvgstyle_p.h
+++ qt6-svg/src/svg/qsvgstyle_p.h
@@ -255,16 +255,6 @@ public:
         return m_paintStyleId;
     }
 
-    void setPaintStyleResolved(bool resolved)
-    {
-        m_paintStyleResolved = resolved;
-    }
-
-    bool isPaintStyleResolved() const
-    {
-        return m_paintStyleResolved;
-    }
-
 private:
     // fill            v 	v 	'inherit' | <Paint.datatype>
     // fill-opacity    v 	v 	'inherit' | <OpacityValue.datatype>
@@ -278,7 +268,6 @@ private:
     qreal m_oldFillOpacity;
 
     QString m_paintStyleId;
-    uint m_paintStyleResolved : 1;
 
     uint m_fillRuleSet : 1;
     uint m_fillOpacitySet : 1;
@@ -479,16 +468,6 @@ public:
         return m_paintStyleId;
     }
 
-    void setPaintStyleResolved(bool resolved)
-    {
-        m_paintStyleResolved = resolved;
-    }
-
-    bool isPaintStyleResolved() const
-    {
-        return m_paintStyleResolved;
-    }
-
     QPen stroke() const
     {
         return m_stroke;
@@ -512,7 +491,6 @@ private:
 
     QSvgPaintStyleProperty *m_style;
     QString m_paintStyleId;
-    uint m_paintStyleResolved : 1;
     uint m_vectorEffect : 1;
     uint m_oldVectorEffect : 1;
 
