Author: Tian Shilin <tianshilin@uniontech.com>
Date:   Mon Mar 16 06:35:21 2026
Subject: Refactor: Create a separate class for paint servers
Upstream: https://codereview.qt-project.org/c/qt/qtsvg/+/721804


---
Index: qt6-svg/src/svg/qsvghandler.cpp
===================================================================
--- qt6-svg.orig/src/svg/qsvghandler.cpp
+++ qt6-svg/src/svg/qsvghandler.cpp
@@ -32,6 +32,7 @@

 #include "float.h"
 #include <cmath>
+#include <memory>

 QT_BEGIN_NAMESPACE

@@ -1007,9 +1008,9 @@ static void parseColor(QSvgNode *,
     }
 }

-static QSvgStyleProperty *styleFromUrl(QSvgNode *node, QStringView url)
+static QSvgPaintServerSharedPtr paintServerFromUrl(QSvgTinyDocument *doc, QStringView url)
 {
-    return node ? node->styleProperty(idFromFuncIRI(url)) : 0;
+    return doc ? doc->paintServer(idFromFuncIRI(url)) : nullptr;
 }

 static void parseBrush(QSvgNode *node,
@@ -1036,11 +1037,9 @@ static void parseBrush(QSvgNode *node,
         if ((!attributes.fill.isEmpty()) && (attributes.fill != QT_INHERIT) ) {
             if (attributes.fill.size() > 3 && attributes.fill.mid(0, 3) == QLatin1String("url")) {
                 QStringView value = attributes.fill;
-                QSvgStyleProperty *style = styleFromUrl(node, value);
-                if (style) {
-                    if (style->type() == QSvgStyleProperty::SOLID_COLOR || style->type() == QSvgStyleProperty::GRADIENT
-                            || style->type() == QSvgStyleProperty::PATTERN)
-                        prop->setFillStyle(reinterpret_cast<QSvgPaintStyleProperty *>(style));
+                QSvgPaintServerSharedPtr paintServer = paintServerFromUrl(handler->document(), value);
+                if (paintServer) {
+                    prop->setPaintServer(paintServer);
                 } else {
                     QString id = idFromFuncIRI(value);
                     prop->setPaintStyleId(id);
@@ -1054,7 +1053,7 @@ static void parseBrush(QSvgNode *node,
                 prop->setBrush(QBrush(Qt::NoBrush));
             }
         }
-        node->appendStyleProperty(prop, attributes.id);
+        node->appendStyleProperty(prop);
     }
 }

@@ -1205,17 +1204,15 @@ static void parsePen(QSvgNode *node,
         //stroke attribute handling
         if ((!attributes.stroke.isEmpty()) && (attributes.stroke != QT_INHERIT) ) {
             if (attributes.stroke.size() > 3 && attributes.stroke.mid(0, 3) == QLatin1String("url")) {
-                    QStringView value = attributes.stroke;
-                    QSvgStyleProperty *style = styleFromUrl(node, value);
-                    if (style) {
-                        if (style->type() == QSvgStyleProperty::SOLID_COLOR || style->type() == QSvgStyleProperty::GRADIENT
-                            || style->type() == QSvgStyleProperty::PATTERN)
-                        prop->setStyle(reinterpret_cast<QSvgPaintStyleProperty *>(style));
-                    } else {
-                        QString id = idFromFuncIRI(value);
-                        prop->setPaintStyleId(id);
-                        handler->pushUnresolvedStyle(prop);
-                    }
+                QStringView value = attributes.stroke;
+                QSvgPaintServerSharedPtr paintServer = paintServerFromUrl(handler->document(), value);
+                if (paintServer) {
+                    prop->setPaintServer(paintServer);
+                } else {
+                    QString id = idFromFuncIRI(value);
+                    prop->setPaintStyleId(id);
+                    handler->pushUnresolvedStyle(prop);
+                }
             } else if (attributes.stroke != QLatin1String("none")) {
                 QColor color;
                 if (resolveColor(attributes.stroke, color, handler))
@@ -1300,7 +1297,7 @@ static void parsePen(QSvgNode *node,
         if (!attributes.strokeOpacity.isEmpty() && attributes.strokeOpacity != QT_INHERIT)
             prop->setOpacity(qMin(qreal(1.0), qMax(qreal(0.0), toDouble(attributes.strokeOpacity))));

-        node->appendStyleProperty(prop, attributes.id);
+        node->appendStyleProperty(prop);
     }
 }

@@ -1436,7 +1433,7 @@ static void parseFont(QSvgNode *node,
            fontStyle->setTextAnchor(Qt::AlignRight);
     }

-    node->appendStyleProperty(fontStyle, attributes.id);
+    node->appendStyleProperty(fontStyle);
 }

 static void parseTransform(QSvgNode *node,
@@ -1448,7 +1445,7 @@ static void parseTransform(QSvgNode *nod
     QTransform matrix = parseTransformationMatrix(attributes.transform.trimmed());

     if (!matrix.isIdentity()) {
-        node->appendStyleProperty(new QSvgTransformStyle(QTransform(matrix)), attributes.id);
+        node->appendStyleProperty(new QSvgTransformStyle(QTransform(matrix)));
     }

 }
@@ -2172,7 +2169,7 @@ static void parseOpacity(QSvgNode *node,

     if (ok) {
         QSvgOpacityStyle *opacity = new QSvgOpacityStyle(qBound(qreal(0.0), op, qreal(1.0)));
-        node->appendStyleProperty(opacity, attributes.id);
+        node->appendStyleProperty(opacity);
     }
 }

@@ -2244,7 +2241,7 @@ static void parseCompOp(QSvgNode *node,

     if (!value.isEmpty()) {
         QSvgCompOpStyle *compop = new QSvgCompOpStyle(svgToQtCompositionMode(value));
-        node->appendStyleProperty(compop, attributes.id);
+        node->appendStyleProperty(compop);
     }
 }

@@ -2354,7 +2351,7 @@ static void parseRenderingHints(QSvgNode
         p->setImageRendering(QSvgQualityStyle::ImageRenderingOptimizeSpeed);
     else if (ir == QLatin1String("optimizeQuality"))
         p->setImageRendering(QSvgQualityStyle::ImageRenderingOptimizeQuality);
-    node->appendStyleProperty(p, attributes.id);
+    node->appendStyleProperty(p);
 }


@@ -2498,7 +2495,7 @@ static bool parseAnimateColorNode(QSvgNo
         (repeatStr == QLatin1String("indefinite")) ? -1 :
             (repeatStr == QLatin1String("")) ? 1 : toDouble(repeatStr));

-    parent->appendStyleProperty(anim, someId(attributes));
+    parent->appendStyleProperty(anim);
     parent->document()->setAnimated(true);
     handler->setAnimPeriod(begin, end);
     return true;
@@ -2611,7 +2608,7 @@ static bool parseAnimateTransformNode(QS
             (repeatStr == QLatin1String("indefinite"))? -1 :
             (repeatStr == QLatin1String(""))? 1 : toDouble(repeatStr));

-    parent->appendStyleProperty(anim, someId(attributes));
+    parent->appendStyleProperty(anim);
     parent->document()->setAnimated(true);
     handler->setAnimPeriod(begin, end);
     return true;
@@ -2926,9 +2923,8 @@ static QSvgNode *createLineNode(QSvgNode
 }


-static void parseBaseGradient(QSvgNode *node,
-                              const QXmlStreamAttributes &attributes,
-                              QSvgGradientStyle *gradProp,
+static void parseBaseGradient(const QXmlStreamAttributes &attributes,
+                              QSvgGradientPaint *gradProp,
                               QSvgHandler *handler)
 {
     QString linkId = idFromIRI(attributes.value(QLatin1String("xlink:href")));
@@ -2946,12 +2942,11 @@ static void parseBaseGradient(QSvgNode *

     QTransform matrix;
     QGradient *grad = gradProp->qgradient();
-    if (node && !linkId.isEmpty()) {
-        QSvgStyleProperty *prop = node->styleProperty(linkId);
-        //qDebug()<<"inherited "<<prop<<" ("<<linkId<<")";
-        if (prop && prop->type() == QSvgStyleProperty::GRADIENT) {
-            QSvgGradientStyle *inherited =
-                static_cast<QSvgGradientStyle*>(prop);
+    if (!linkId.isEmpty()) {
+        QSvgPaintServerSharedPtr paintServer = handler->document()->paintServer(linkId);
+        if (paintServer && paintServer->type() == QSvgPaintServer::Type::Gradient) {
+            QSvgGradientPaint *inherited =
+                static_cast<QSvgGradientPaint*>(paintServer.get());
             if (!inherited->stopLink().isEmpty()) {
                 gradProp->setStopLink(inherited->stopLink(), handler->document());
             } else {
@@ -2987,9 +2982,8 @@ static void parseBaseGradient(QSvgNode *
     }
 }

-static QSvgStyleProperty *createLinearGradientNode(QSvgNode *node,
-                                                   const QXmlStreamAttributes &attributes,
-                                                   QSvgHandler *handler)
+static QSvgPaintServerSharedPtr createLinearGradientNode(const QXmlStreamAttributes &attributes,
+                                                         QSvgHandler *handler)
 {
     const QStringView x1 = attributes.value(QLatin1String("x1"));
     const QStringView y1 = attributes.value(QLatin1String("y1"));
@@ -3010,17 +3004,13 @@ static QSvgStyleProperty *createLinearGr
     if (!y2.isEmpty())
         ny2 =  convertToNumber(y2, handler);

-    QSvgNode *itr = node;
-    while (itr && itr->type() != QSvgNode::Doc) {
-        itr = itr->parent();
-    }
-
-    QLinearGradient *grad = new QLinearGradient(nx1, ny1, nx2, ny2);
+    auto grad = std::make_unique<QLinearGradient>(nx1, ny1, nx2, ny2);
     grad->setInterpolationMode(QGradient::ComponentInterpolation);
-    QSvgGradientStyle *prop = new QSvgGradientStyle(grad);
-    parseBaseGradient(node, attributes, prop, handler);

-    return prop;
+    QSvgGradientPaintSharedPtr paintServer = std::make_shared<QSvgGradientPaint>(std::move(grad));
+    parseBaseGradient(attributes, paintServer.get(), handler);
+
+    return paintServer;
 }

 static bool parseMetadataNode(QSvgNode *parent,
@@ -3770,9 +3760,8 @@ static bool parsePrefetchNode(QSvgNode *
     return true;
 }

-static QSvgStyleProperty *createRadialGradientNode(QSvgNode *node,
-                                                   const QXmlStreamAttributes &attributes,
-                                                   QSvgHandler *handler)
+static QSvgPaintServerSharedPtr createRadialGradientNode(const QXmlStreamAttributes &attributes,
+                                                         QSvgHandler *handler)
 {
     const QStringView cx = attributes.value(QLatin1String("cx"));
     const QStringView cy = attributes.value(QLatin1String("cy"));
@@ -3800,13 +3789,13 @@ static QSvgStyleProperty *createRadialGr
     if (!fy.isEmpty())
         nfy = toDouble(fy);

-    QRadialGradient *grad = new QRadialGradient(ncx, ncy, nr, nfx, nfy, 0);
+    auto grad = std::make_unique<QRadialGradient>(ncx, ncy, nr, nfx, nfy, 0);
     grad->setInterpolationMode(QGradient::ComponentInterpolation);

-    QSvgGradientStyle *prop = new QSvgGradientStyle(grad);
-    parseBaseGradient(node, attributes, prop, handler);
+    QSvgGradientPaintSharedPtr paintServer = std::make_shared<QSvgGradientPaint>(std::move(grad));
+    parseBaseGradient(attributes, paintServer.get(), handler);

-    return prop;
+    return paintServer;
 }

 static QSvgNode *createRectNode(QSvgNode *parent,
@@ -3876,11 +3865,10 @@ static bool parseSetNode(QSvgNode *paren
     return true;
 }

-static QSvgStyleProperty *createSolidColorNode(QSvgNode *parent,
-                                               const QXmlStreamAttributes &attributes,
-                                               QSvgHandler *handler)
+static QSvgPaintServerSharedPtr createSolidColorNode(const QXmlStreamAttributes &attributes,
+                                                     QSvgHandler *handler)
 {
-    Q_UNUSED(parent); Q_UNUSED(attributes);
+    Q_UNUSED(attributes);
     QStringView solidColorStr = attributes.value(QLatin1String("solid-color"));
     QStringView solidOpacityStr = attributes.value(QLatin1String("solid-opacity"));

@@ -3889,16 +3877,16 @@ static QSvgStyleProperty *createSolidCol

     QColor color;
     if (!constructColor(solidColorStr, solidOpacityStr, color, handler))
-        return 0;
-    QSvgSolidColorStyle *style = new QSvgSolidColorStyle(color);
-    return style;
+        return nullptr;
+    std::shared_ptr<QSvgSolidColorPaint> paintServer = std::make_shared<QSvgSolidColorPaint>(color);
+    return paintServer;
 }

-static bool parseStopNode(QSvgStyleProperty *parent,
+static bool parseStopNode(QSvgPaintServer *paintServer,
                           const QXmlStreamAttributes &attributes,
                           QSvgHandler *handler)
 {
-    if (parent->type() != QSvgStyleProperty::GRADIENT)
+    if (paintServer->type() != QSvgPaintServer::Type::Gradient)
         return false;
     QString nodeIdStr     = someId(attributes);
     QString xmlClassStr   = attributes.value(QLatin1String("class")).toString();
@@ -3920,8 +3908,7 @@ static bool parseStopNode(QSvgStylePrope

     QSvgAttributes attrs(xmlAttr, handler);

-    QSvgGradientStyle *style =
-        static_cast<QSvgGradientStyle*>(parent);
+    QSvgGradientPaint *style = static_cast<QSvgGradientPaint*>(paintServer);
     QStringView colorStr    = attrs.stopColor;
     QColor color;

@@ -4144,8 +4131,8 @@ static QSvgNode *createPatternNode(QSvgN
     QSvgPattern *node = new QSvgPattern(parent, patternRectF, viewBox, nPatternContentUnits, matrix);

     // Create a style node for the Pattern.
-    QSvgPatternStyle *prop = new QSvgPatternStyle(node);
-    node->appendStyleProperty(prop, someId(attributes));
+    QSvgPaintServerSharedPtr prop = std::make_shared<QSvgPatternPaint>(node);
+    handler->document()->addPaintServer(prop, someId(attributes));

     return node;
 }
@@ -4448,15 +4435,6 @@ static StyleFactoryMethod findStyleFacto
     case 'f':
         if (ref == QLatin1String("ont")) return createFontNode;
         break;
-    case 'l':
-        if (ref == QLatin1String("inearGradient")) return createLinearGradientNode;
-        break;
-    case 'r':
-        if (ref == QLatin1String("adialGradient")) return createRadialGradientNode;
-        break;
-    case 's':
-        if (ref == QLatin1String("olidColor")) return createSolidColorNode;
-        break;
     default:
         break;
     }
@@ -4486,6 +4464,48 @@ static StyleParseMethod findStyleUtilFac
     case 'm':
         if (ref == QLatin1String("issing-glyph")) return parseMissingGlyphNode;
         break;
+    default:
+        break;
+    }
+    return 0;
+}
+
+typedef QSvgPaintServerSharedPtr (*PaintServerFactoryMethod)(const QXmlStreamAttributes &,
+                                                                     QSvgHandler *);
+
+static PaintServerFactoryMethod findPaintServerFactoryMethod(const QStringView name)
+{
+    if (name.isEmpty())
+        return nullptr;
+
+    QStringView ref = name.sliced(1);
+    switch (name.at(0).unicode()) {
+    case 'l':
+        if (ref == QLatin1String("inearGradient")) return createLinearGradientNode;
+        break;
+    case 'r':
+        if (ref == QLatin1String("adialGradient")) return createRadialGradientNode;
+        break;
+    case 's':
+        if (ref == QLatin1String("olidColor")) return createSolidColorNode;
+        break;
+    default:
+        break;
+    }
+    return nullptr;
+}
+
+typedef bool (*PaintServerParseMethod)(QSvgPaintServer *,
+                                 const QXmlStreamAttributes &,
+                                 QSvgHandler *);
+
+static PaintServerParseMethod findPaintServerUtilFactoryMethod(const QStringView name)
+{
+    if (name.isEmpty())
+        return 0;
+
+    QStringView ref = name.sliced(1);
+    switch (name.at(0).unicode()) {
     case 's':
         if (ref == QLatin1String("top")) return parseStopNode;
         break;
@@ -4523,6 +4543,7 @@ void QSvgHandler::init()
 {
     m_doc = 0;
     m_style = 0;
+    m_paintServer.reset();
     m_animEnd = 0;
     m_defaultCoords = LT_PX;
     m_defaultPen = QPen(Qt::black, 1, Qt::SolidLine, Qt::FlatCap, Qt::SvgMiterJoin);
@@ -4534,16 +4555,18 @@ static bool detectPatternCycles(const QS
 {
     QSvgFillStyle *fillStyle = static_cast<QSvgFillStyle*>
         (node->styleProperty(QSvgStyleProperty::FILL));
-    if (fillStyle && fillStyle->style() && fillStyle->style()->type() == QSvgStyleProperty::PATTERN) {
-        QSvgPatternStyle *patternStyle = static_cast<QSvgPatternStyle *>(fillStyle->style());
+    if (fillStyle && fillStyle->paintServer()
+        && fillStyle->paintServer()->type() == QSvgPaintServer::Type::Pattern) {
+        QSvgPatternPaint *patternStyle = static_cast<QSvgPatternPaint *>(fillStyle->paintServer());
         if (active.contains(patternStyle->patternNode()))
             return true;
     }

     QSvgStrokeStyle *strokeStyle = static_cast<QSvgStrokeStyle*>
         (node->styleProperty(QSvgStyleProperty::STROKE));
-    if (strokeStyle && strokeStyle->style() && strokeStyle->style()->type() == QSvgStyleProperty::PATTERN) {
-        QSvgPatternStyle *patternStyle = static_cast<QSvgPatternStyle *>(strokeStyle->style());
+    if (strokeStyle && strokeStyle->paintServer()
+        && strokeStyle->paintServer()->type() == QSvgPaintServer::Type::Pattern) {
+        QSvgPatternPaint *patternStyle = static_cast<QSvgPatternPaint *>(strokeStyle->paintServer());
         if (active.contains(patternStyle->patternNode()))
             return true;
     }
@@ -4823,7 +4846,16 @@ bool QSvgHandler::startElement(const QSt
         QSvgStyleProperty *prop = method(m_nodes.top(), attributes, this);
         if (prop) {
             m_style = prop;
-            m_nodes.top()->appendStyleProperty(prop, someId(attributes));
+            m_nodes.top()->appendStyleProperty(prop);
+        } else {
+            const QByteArray msg = QByteArrayLiteral("Could not parse node: ") + localName.toLocal8Bit();
+            qCWarning(lcSvgHandler, "%s", prefixMessage(msg, xml).constData());
+        }
+    } else if (PaintServerFactoryMethod method = findPaintServerFactoryMethod(localName)) {
+        QSvgPaintServerSharedPtr paintServer = method(attributes, this);
+        if (paintServer) {
+            m_paintServer = paintServer;
+            m_doc->addPaintServer(paintServer, someId(attributes));
         } else {
             const QByteArray msg = QByteArrayLiteral("Could not parse node: ") + localName.toLocal8Bit();
             qCWarning(lcSvgHandler, "%s", prefixMessage(msg, xml).constData());
@@ -4833,6 +4865,11 @@ bool QSvgHandler::startElement(const QSt
             if (!method(m_style, attributes, this))
                 qCWarning(lcSvgHandler, "%s", msgProblemParsing(localName, xml).constData());
         }
+    } else if (PaintServerParseMethod method = findPaintServerUtilFactoryMethod(localName)) {
+        if (m_paintServer) {
+            if (!method(m_paintServer.get(), attributes, this))
+                qCWarning(lcSvgHandler, "%s", msgProblemParsing(localName, xml).constData());
+        }
     } else {
         qCDebug(lcSvgHandler) << "Skipping unknown element" << localName;
         m_skipNodes.push(Unknown);
@@ -4875,6 +4912,8 @@ bool QSvgHandler::endElement(const QStri
         m_nodes.pop();
     else if (m_style && !m_skipNodes.isEmpty() && m_skipNodes.top() != Style)
         m_style = 0;
+    else if (m_paintServer && !m_skipNodes.isEmpty() && m_skipNodes.top() != Style)
+        m_paintServer.reset();

     return ((localName == QLatin1String("svg")) && (node != Doc));
 }
@@ -4885,9 +4924,9 @@ void QSvgHandler::resolvePaintServers()
         if (prop->type() == QSvgStyleProperty::FILL) {
             QSvgFillStyle *fill = static_cast<QSvgFillStyle *>(prop);
             QString id = fill->paintStyleId();
-            QSvgPaintStyleProperty *style = m_doc->namedStyle(id);
-            if (style) {
-                fill->setFillStyle(style);
+            QSvgPaintServerSharedPtr paintServer = m_doc->paintServer(id);
+            if (paintServer) {
+                fill->setPaintServer(paintServer);
             } else {
                 qCWarning(lcSvgHandler, "%s", msgCouldNotResolveProperty(id, xml).constData());
                 fill->setBrush(Qt::NoBrush);
@@ -4895,9 +4934,9 @@ void QSvgHandler::resolvePaintServers()
         } else if (prop->type() == QSvgStyleProperty::STROKE) {
             QSvgStrokeStyle *stroke = static_cast<QSvgStrokeStyle *>(prop);
             QString id = stroke->paintStyleId();
-            QSvgPaintStyleProperty *style = m_doc->namedStyle(id);
-            if (style) {
-                stroke->setStyle(style);
+            QSvgPaintServerSharedPtr paintServer = m_doc->paintServer(id);
+            if (paintServer) {
+                stroke->setPaintServer(paintServer);
             } else {
                 qCWarning(lcSvgHandler, "%s", msgCouldNotResolveProperty(id, xml).constData());
                 stroke->setStroke(Qt::NoBrush);
Index: qt6-svg/src/svg/qsvgnode.cpp
===================================================================
--- qt6-svg.orig/src/svg/qsvgnode.cpp
+++ qt6-svg/src/svg/qsvgnode.cpp
@@ -178,10 +178,8 @@ bool QSvgNode::isDescendantOf(const QSvg
     return false;
 }

-void QSvgNode::appendStyleProperty(QSvgStyleProperty *prop, const QString &id)
+void QSvgNode::appendStyleProperty(QSvgStyleProperty *prop)
 {
-    //qDebug()<<"appending "<<prop->type()<< " ("<< id <<") "<<"to "<<this<<this->type();
-    QSvgTinyDocument *doc;
     switch (prop->type()) {
     case QSvgStyleProperty::QUALITY:
         m_style.quality = static_cast<QSvgQualityStyle*>(prop);
@@ -198,24 +196,6 @@ void QSvgNode::appendStyleProperty(QSvgS
     case QSvgStyleProperty::STROKE:
         m_style.stroke = static_cast<QSvgStrokeStyle*>(prop);
         break;
-    case QSvgStyleProperty::SOLID_COLOR:
-        m_style.solidColor = static_cast<QSvgSolidColorStyle*>(prop);
-        doc = document();
-        if (doc && !id.isEmpty())
-            doc->addNamedStyle(id, m_style.solidColor);
-        break;
-    case QSvgStyleProperty::GRADIENT:
-        m_style.gradient = static_cast<QSvgGradientStyle*>(prop);
-        doc = document();
-        if (doc && !id.isEmpty())
-            doc->addNamedStyle(id, m_style.gradient);
-        break;
-    case QSvgStyleProperty::PATTERN:
-        m_style.pattern = static_cast<QSvgPatternStyle*>(prop);
-        doc = document();
-        if (doc && !id.isEmpty())
-            doc->addNamedStyle(id, m_style.pattern);
-        break;
     case QSvgStyleProperty::TRANSFORM:
         m_style.transform = static_cast<QSvgTransformStyle*>(prop);
         break;
@@ -296,18 +276,6 @@ QSvgStyleProperty * QSvgNode::stylePrope
             if (node->m_style.stroke)
                 return node->m_style.stroke;
             break;
-        case QSvgStyleProperty::SOLID_COLOR:
-            if (node->m_style.solidColor)
-                return node->m_style.solidColor;
-            break;
-        case QSvgStyleProperty::GRADIENT:
-            if (node->m_style.gradient)
-                return node->m_style.gradient;
-            break;
-        case QSvgStyleProperty::PATTERN:
-            if (node->m_style.pattern)
-                return node->m_style.pattern;
-            break;
         case QSvgStyleProperty::TRANSFORM:
             if (node->m_style.transform)
                 return node->m_style.transform;
@@ -337,12 +305,6 @@ QSvgStyleProperty * QSvgNode::stylePrope
     return 0;
 }

-QSvgPaintStyleProperty * QSvgNode::styleProperty(const QString &id) const
-{
-    QSvgTinyDocument *doc = document();
-    return doc ? doc->namedStyle(id) : 0;
-}
-
 QRectF QSvgNode::internalFastBounds(QPainter *p, QSvgExtraStates &states) const
 {
     return internalBounds(p, states);
Index: qt6-svg/src/svg/qsvgnode_p.h
===================================================================
--- qt6-svg.orig/src/svg/qsvgnode_p.h
+++ qt6-svg/src/svg/qsvgnode_p.h
@@ -100,14 +100,12 @@ public:
     QSvgNode *parent() const;
     bool isDescendantOf(const QSvgNode *parent) const;

-    void appendStyleProperty(QSvgStyleProperty *prop, const QString &id);
+    void appendStyleProperty(QSvgStyleProperty *prop);
     void applyStyle(QPainter *p, QSvgExtraStates &states) const;
     void applyStyleRecursive(QPainter *p, QSvgExtraStates &states) const;
     void revertStyle(QPainter *p, QSvgExtraStates &states) const;
     void revertStyleRecursive(QPainter *p, QSvgExtraStates &states) const;
     QSvgStyleProperty *styleProperty(QSvgStyleProperty::Type type) const;
-    QSvgPaintStyleProperty *styleProperty(const QString &id) const;
-
     QSvgTinyDocument *document() const;

     virtual Type type() const = 0;
Index: qt6-svg/src/svg/qsvgpaintserver.cpp
===================================================================
--- /dev/null
+++ qt6-svg/src/svg/qsvgpaintserver.cpp
@@ -0,0 +1,90 @@
+// Copyright (C) 2026 The Qt Company Ltd.
+// SPDX-License-Identifier: LicenseRef-Qt-Commercial OR LGPL-3.0-only OR GPL-2.0-only OR GPL-3.0-only
+// Qt-Security score:significant reason:default
+
+#include "qsvgpaintserver_p.h"
+#include <QtSvg/private/qsvgnode_p.h>
+#include <QtSvg/private/qsvgtinydocument_p.h>
+
+QT_BEGIN_NAMESPACE
+
+QSvgPaintServer::~QSvgPaintServer()
+    = default;
+
+QSvgSolidColorPaint::QSvgSolidColorPaint(const QColor &color)
+    : m_solidColor(color)
+{
+}
+
+QSvgSolidColorPaint::~QSvgSolidColorPaint()
+    = default;
+
+QSvgGradientPaint::QSvgGradientPaint(std::unique_ptr<QGradient> grad)
+    : m_gradient(std::move(grad))
+{
+}
+
+QSvgGradientPaint::~QSvgGradientPaint()
+    = default;
+
+QBrush QSvgGradientPaint::brush(QPainter *, const QSvgNode *, QSvgExtraStates &)
+{
+    if (!m_link.isEmpty())
+        resolveStops();
+
+    // If the gradient is marked as empty, insert transparent black
+    if (!m_gradientStopsSet) {
+        m_gradient->setStops(QGradientStops() << QGradientStop(0.0, QColor(0, 0, 0, 0)));
+        m_gradientStopsSet = true;
+    }
+
+    QBrush b(*m_gradient);
+
+    if (!m_transform.isIdentity())
+        b.setTransform(m_transform);
+
+    return b;
+}
+
+void QSvgGradientPaint::resolveStops()
+{
+    QSet<QString> visited;
+    resolveStops_helper(visited);
+}
+
+void QSvgGradientPaint::resolveStops_helper(QSet<QString> &visited)
+{
+    if (!m_link.isEmpty() && m_doc) {
+        QSvgPaintServerSharedPtr paintServer = m_doc->paintServer(m_link);
+        if (paintServer && !visited.contains(m_link)) {
+            visited.insert(m_link);
+            if (paintServer->type() == QSvgPaintServer::Type::Gradient) {
+                QSvgGradientPaint *st =
+                    static_cast<QSvgGradientPaint*>(paintServer.get());
+                st->resolveStops_helper(visited);
+                m_gradient->setStops(st->qgradient()->stops());
+                m_gradientStopsSet = st->gradientStopsSet();
+            }
+        } else {
+            qWarning("Could not resolve property : %s", qPrintable(m_link));
+        }
+        m_link.clear();
+    }
+}
+
+QSvgPatternPaint::QSvgPatternPaint(QSvgPattern *pattern)
+    : m_pattern(pattern)
+{
+}
+
+QSvgPatternPaint::~QSvgPatternPaint()
+    = default;
+
+QBrush QSvgPatternPaint::brush(QPainter *p, const QSvgNode *node, QSvgExtraStates &states)
+{
+    QBrush b(m_pattern->patternImage(p, states, node));
+    b.setTransform(m_pattern->appliedTransform());
+    return b;
+}
+
+QT_END_NAMESPACE
Index: qt6-svg/src/svg/qsvgpaintserver_p.h
===================================================================
--- /dev/null
+++ qt6-svg/src/svg/qsvgpaintserver_p.h
@@ -0,0 +1,192 @@
+// Copyright (C) 2026 The Qt Company Ltd.
+// SPDX-License-Identifier: LicenseRef-Qt-Commercial OR LGPL-3.0-only OR GPL-2.0-only OR GPL-3.0-only
+// Qt-Security score:significant reason:default
+
+
+#ifndef QSVGPAINTSERVER_P_H
+#define QSVGPAINTSERVER_P_H
+
+//
+//  W A R N I N G
+//  -------------
+//
+// This file is not part of the Qt API.  It exists purely as an
+// implementation detail.  This header file may change from version to
+// version without notice, or even be removed.
+//
+// We mean it.
+//
+
+#include <memory>
+#include <QtCore/qrect.h>
+#include <QtCore/qset.h>
+#include <QtCore/qstring.h>
+#include <QtGui/qcolor.h>
+#include <QtGui/qbrush.h>
+#include <QtGui/qpainter.h>
+#include <QtGui/qpen.h>
+#include <QtGui/qtransform.h>
+#include "qtsvgexports.h"
+
+QT_BEGIN_NAMESPACE
+
+class QSvgNode;
+class QSvgTinyDocument;
+class QSvgPattern;
+struct QSvgExtraStates;
+
+class Q_SVG_EXPORT QSvgPaintServer
+{
+    Q_DISABLE_COPY_MOVE(QSvgPaintServer)
+public:
+    enum class Type : qint8 {
+        SolidColor,
+        Gradient,
+        Pattern,
+    };
+public:
+    QSvgPaintServer() = default;
+    virtual ~QSvgPaintServer();
+    virtual QBrush brush(QPainter *p, const QSvgNode *node, QSvgExtraStates &states) = 0;
+    virtual Type type() const = 0;
+};
+
+class Q_SVG_EXPORT QSvgSolidColorPaint : public QSvgPaintServer
+{
+public:
+    QSvgSolidColorPaint(const QColor &color);
+    ~QSvgSolidColorPaint() override;
+    Type type() const override;
+
+    const QColor & qcolor() const;
+
+    QBrush brush(QPainter *, const QSvgNode *, QSvgExtraStates &) override;
+
+private:
+    // solid-color       v 	x 	'inherit' | <SVGColor.datatype>
+    // solid-opacity     v 	x 	'inherit' | <OpacityValue.datatype>
+    QColor m_solidColor;
+
+    QBrush m_oldFill;
+    QPen m_oldStroke;
+};
+
+inline QSvgPaintServer::Type QSvgSolidColorPaint::type() const
+{
+    return Type::SolidColor;
+}
+
+inline const QColor &QSvgSolidColorPaint::qcolor() const
+{
+    return m_solidColor;
+}
+
+inline QBrush QSvgSolidColorPaint::brush(QPainter *, const QSvgNode *, QSvgExtraStates &)
+{
+    return m_solidColor;
+}
+
+
+class Q_SVG_EXPORT QSvgGradientPaint : public QSvgPaintServer
+{
+public:
+    QSvgGradientPaint(std::unique_ptr<QGradient> grad);
+    ~QSvgGradientPaint() override;
+    Type type() const override;
+
+    void setStopLink(const QString &link, QSvgTinyDocument *doc);
+    QString stopLink() const;
+
+    void setTransform(const QTransform &transform);
+    QTransform qtransform() const;
+
+    QGradient *qgradient();
+
+    bool gradientStopsSet() const;
+
+    void setGradientStopsSet(bool set);
+
+    QBrush brush(QPainter *, const QSvgNode *, QSvgExtraStates &) override;
+private:
+    void resolveStops();
+    void resolveStops_helper(QSet<QString> &visited);
+private:
+    std::unique_ptr<QGradient> m_gradient;
+    QTransform m_transform;
+
+    QSvgTinyDocument *m_doc{nullptr};
+    QString m_link;
+    bool m_gradientStopsSet{false};
+};
+
+inline QSvgPaintServer::Type QSvgGradientPaint::type() const
+{
+    return Type::Gradient;
+}
+
+inline void QSvgGradientPaint::setStopLink(const QString &link, QSvgTinyDocument *doc)
+{
+    m_link = link;
+    m_doc  = doc;
+}
+
+inline QString QSvgGradientPaint::stopLink() const
+{
+    return m_link;
+}
+
+inline void QSvgGradientPaint::setTransform(const QTransform &transform)
+{
+    m_transform = transform;
+}
+
+inline QTransform QSvgGradientPaint::qtransform() const
+{
+    return m_transform;
+}
+
+inline QGradient *QSvgGradientPaint::qgradient()
+{
+    return m_gradient.get();
+}
+
+inline void QSvgGradientPaint::setGradientStopsSet(bool set)
+{
+    m_gradientStopsSet = set;
+}
+
+inline bool QSvgGradientPaint::gradientStopsSet() const
+{
+    return m_gradientStopsSet;
+}
+
+class Q_SVG_EXPORT QSvgPatternPaint : public QSvgPaintServer
+{
+public:
+    QSvgPatternPaint(QSvgPattern *pattern);
+    ~QSvgPatternPaint() override;
+    Type type() const override;
+
+    QBrush brush(QPainter *, const QSvgNode *, QSvgExtraStates &) override;
+    QSvgPattern *patternNode();
+private:
+    QSvgPattern *m_pattern;
+    QRectF m_parentBound;
+};
+
+inline QSvgPaintServer::Type QSvgPatternPaint::type() const
+{
+    return Type::Pattern;
+}
+
+inline QSvgPattern *QSvgPatternPaint::patternNode()
+{
+    return m_pattern;
+}
+
+using QSvgPaintServerSharedPtr = std::shared_ptr<QSvgPaintServer>;
+using QSvgGradientPaintSharedPtr = std::shared_ptr<QSvgGradientPaint>;
+
+QT_END_NAMESPACE
+
+#endif // QSVGPAINTSERVER_P_H
Index: qt6-svg/src/svg/qsvgstructure.cpp
===================================================================
--- qt6-svg.orig/src/svg/qsvgstructure.cpp
+++ qt6-svg/src/svg/qsvgstructure.cpp
@@ -194,7 +194,7 @@ QSvgMarker::QSvgMarker(QSvgNode *parent,
     // apply the svg standard style
     QSvgFillStyle *fillProp = new QSvgFillStyle();
     fillProp->setBrush(Qt::black);
-    appendStyleProperty(fillProp, QStringLiteral(""));
+    appendStyleProperty(fillProp);

     QSvgStrokeStyle *strokeProp = new QSvgStrokeStyle();
     strokeProp->setMiterLimit(4);
@@ -202,7 +202,7 @@ QSvgMarker::QSvgMarker(QSvgNode *parent,
     strokeProp->setLineCap(Qt::FlatCap);
     strokeProp->setLineJoin(Qt::SvgMiterJoin);
     strokeProp->setStroke(Qt::NoBrush);
-    appendStyleProperty(strokeProp, QStringLiteral(""));
+    appendStyleProperty(strokeProp);
 }

 QSvgFilterContainer::QSvgFilterContainer(QSvgNode *parent, const QSvgRectF &bounds,
Index: qt6-svg/src/svg/qsvgstyle_p.h
===================================================================
--- qt6-svg.orig/src/svg/qsvgstyle_p.h
+++ qt6-svg/src/svg/qsvgstyle_p.h
@@ -20,10 +20,10 @@
 #include "QtGui/qpen.h"
 #include "QtGui/qbrush.h"
 #include "QtGui/qtransform.h"
-#include "QtGui/qcolor.h"
 #include "QtGui/qfont.h"
 #include <qdebug.h>
 #include "qtsvgglobal_p.h"
+#include <QtSvg/private/qsvgpaintserver_p.h>

 QT_BEGIN_NAMESPACE

@@ -132,9 +132,6 @@ public:
         VIEWPORT_FILL,
         FONT,
         STROKE,
-        SOLID_COLOR,
-        GRADIENT,
-        PATTERN,
         TRANSFORM,
         ANIMATE_TRANSFORM,
         ANIMATE_COLOR,
@@ -149,14 +146,6 @@ public:
     bool isDefault() const { return false; } // [not virtual since called from templated class]
 };

-class Q_SVG_EXPORT QSvgPaintStyleProperty : public QSvgStyleProperty
-{
-public:
-    virtual QBrush brush(QPainter *p, const QSvgNode *node, QSvgExtraStates &states) = 0;
-    void apply(QPainter *p, const QSvgNode *node, QSvgExtraStates &states) override;
-    void revert(QPainter *p, QSvgExtraStates &states) override;
-};
-
 class Q_SVG_EXPORT QSvgQualityStyle : public QSvgStyleProperty
 {
 public:
@@ -224,7 +213,7 @@ public:

     void setFillRule(Qt::FillRule f);
     void setFillOpacity(qreal opacity);
-    void setFillStyle(QSvgPaintStyleProperty* style);
+    void setPaintServer(QSvgPaintServerSharedPtr paintServer);
     void setBrush(QBrush brush);

     const QBrush & qbrush() const
@@ -242,9 +231,9 @@ public:
         return m_fillRule;
     }

-    QSvgPaintStyleProperty* style() const
+    QSvgPaintServer *paintServer() const
     {
-        return m_style;
+        return m_paintServer.get();
     }

     void setPaintStyleId(const QString &Id)
@@ -262,7 +251,7 @@ private:
     // fill-opacity    v 	v 	'inherit' | <OpacityValue.datatype>
     QBrush m_fill;
     QBrush m_oldFill;
-    QSvgPaintStyleProperty *m_style;
+    QSvgPaintServerSharedPtr m_paintServer;

     Qt::FillRule m_fillRule;
     Qt::FillRule m_oldFillRule;
@@ -389,13 +378,13 @@ public:
     void setStroke(QBrush brush)
     {
         m_stroke.setBrush(brush);
-        m_style = nullptr;
+        m_paintServer.reset();
         m_strokeSet = 1;
     }

-    void setStyle(QSvgPaintStyleProperty *style)
+    void setPaintServer(QSvgPaintServerSharedPtr paintServer)
     {
-        m_style = style;
+        m_paintServer = paintServer;
         m_strokeSet = 1;
     }

@@ -455,9 +444,9 @@ public:
         m_vectorEffectSet = 1;
     }

-    QSvgPaintStyleProperty* style() const
+    QSvgPaintServer *paintServer() const
     {
-        return m_style;
+        return m_paintServer.get();
     }

     void setPaintStyleId(const QString &Id)
@@ -491,7 +480,7 @@ private:
     qreal m_strokeDashOffset;
     qreal m_oldStrokeDashOffset;

-    QSvgPaintStyleProperty *m_style;
+    QSvgPaintServerSharedPtr m_paintServer;
     QString m_paintStyleId;
     uint m_vectorEffect : 1;
     uint m_oldVectorEffect : 1;
@@ -507,90 +496,6 @@ private:
     uint m_vectorEffectSet : 1;
 };

-class Q_SVG_EXPORT QSvgSolidColorStyle : public QSvgPaintStyleProperty
-{
-public:
-    QSvgSolidColorStyle(const QColor &color);
-    Type type() const override;
-
-    const QColor & qcolor() const
-    {
-        return m_solidColor;
-    }
-
-    QBrush brush(QPainter *, const QSvgNode *, QSvgExtraStates &) override
-    {
-        return m_solidColor;
-    }
-
-private:
-    // solid-color       v 	x 	'inherit' | <SVGColor.datatype>
-    // solid-opacity     v 	x 	'inherit' | <OpacityValue.datatype>
-    QColor m_solidColor;
-
-    QBrush m_oldFill;
-    QPen   m_oldStroke;
-};
-
-class Q_SVG_EXPORT QSvgGradientStyle : public QSvgPaintStyleProperty
-{
-public:
-    QSvgGradientStyle(QGradient *grad);
-    ~QSvgGradientStyle() { delete m_gradient; }
-    Type type() const override;
-
-    void setStopLink(const QString &link, QSvgTinyDocument *doc);
-    QString stopLink() const { return m_link; }
-    void resolveStops();
-    void resolveStops_helper(QStringList *visited);
-
-    void setTransform(const QTransform &transform);
-    QTransform qtransform() const
-    {
-        return m_transform;
-    }
-
-    QGradient *qgradient() const
-    {
-        return m_gradient;
-    }
-
-    bool gradientStopsSet() const
-    {
-        return m_gradientStopsSet;
-    }
-
-    void setGradientStopsSet(bool set)
-    {
-        m_gradientStopsSet = set;
-    }
-
-    QBrush brush(QPainter *, const QSvgNode *, QSvgExtraStates &) override;
-private:
-    QGradient      *m_gradient;
-    QTransform m_transform;
-
-    QSvgTinyDocument *m_doc;
-    QString           m_link;
-    bool m_gradientStopsSet;
-};
-
-class Q_SVG_EXPORT QSvgPatternStyle : public QSvgPaintStyleProperty
-{
-public:
-    QSvgPatternStyle(QSvgPattern *pattern);
-    ~QSvgPatternStyle() = default;
-    Type type() const override;
-
-    QBrush brush(QPainter *, const QSvgNode *, QSvgExtraStates &) override;
-    QSvgPattern *patternNode() { return m_pattern; }
-private:
-    QSvgPattern *m_pattern;
-    QImage m_patternImage;
-    QRectF m_parentBound;
-};
-
-
 class Q_SVG_EXPORT QSvgTransformStyle : public QSvgStyleProperty
 {
 public:
@@ -736,9 +641,6 @@ public:
           viewportFill(0),
           font(0),
           stroke(0),
-          solidColor(0),
-          gradient(0),
-          pattern(0),
           transform(0),
           animateColor(0),
           opacity(0),
@@ -753,9 +655,6 @@ public:
     QSvgRefCounter<QSvgViewportFillStyle> viewportFill;
     QSvgRefCounter<QSvgFontStyle>         font;
     QSvgRefCounter<QSvgStrokeStyle>       stroke;
-    QSvgRefCounter<QSvgSolidColorStyle>   solidColor;
-    QSvgRefCounter<QSvgGradientStyle>     gradient;
-    QSvgRefCounter<QSvgPatternStyle>      pattern;
     QSvgRefCounter<QSvgTransformStyle>    transform;
     QSvgRefCounter<QSvgAnimateColor>      animateColor;
     QList<QSvgRefCounter<QSvgAnimateTransform> >   animateTransforms;
Index: qt6-svg/src/svg/CMakeLists.txt
===================================================================
--- qt6-svg.orig/src/svg/CMakeLists.txt
+++ qt6-svg/src/svg/CMakeLists.txt
@@ -23,6 +23,7 @@ qt_internal_add_module(Svg
         qsvgstructure.cpp qsvgstructure_p.h
         qsvgfilter.cpp qsvgfilter_p.h
         qsvgstyle.cpp qsvgstyle_p.h
+        qsvgpaintserver.cpp qsvgpaintserver_p.h
         qsvgtinydocument.cpp qsvgtinydocument_p.h
         qsvghelper_p.h
         qsvgdebug.cpp
Index: qt6-svg/src/svg/qsvghandler_p.h
===================================================================
--- qt6-svg.orig/src/svg/qsvghandler_p.h
+++ qt6-svg/src/svg/qsvghandler_p.h
@@ -138,6 +138,7 @@ private:
     QStack<QSvgText::WhitespaceMode> m_whitespaceMode;

     QSvgRefCounter<QSvgStyleProperty> m_style;
+    QSvgPaintServerSharedPtr m_paintServer;

     LengthType m_defaultCoords;

Index: qt6-svg/src/svg/qsvgstyle.cpp
===================================================================
--- qt6-svg.orig/src/svg/qsvgstyle.cpp
+++ qt6-svg/src/svg/qsvgstyle.cpp
@@ -34,17 +34,6 @@ QSvgStyleProperty::~QSvgStyleProperty()
 {
 }

-void QSvgPaintStyleProperty::apply(QPainter *, const QSvgNode *, QSvgExtraStates &)
-{
-    Q_ASSERT(!"This should not be called!");
-}
-
-void QSvgPaintStyleProperty::revert(QPainter *, QSvgExtraStates &)
-{
-    Q_ASSERT(!"This should not be called!");
-}
-
-
 QSvgQualityStyle::QSvgQualityStyle(int color)
     : m_imageRendering(QSvgQualityStyle::ImageRenderingAuto)
     , m_oldImageRendering(QSvgQualityStyle::ImageRenderingAuto)
@@ -89,8 +78,7 @@ void QSvgQualityStyle::revert(QPainter *
 }

 QSvgFillStyle::QSvgFillStyle()
-    : m_style(0)
-    , m_fillRule(Qt::WindingFill)
+    : m_fillRule(Qt::WindingFill)
     , m_oldFillRule(Qt::WindingFill)
     , m_fillOpacity(1.0)
     , m_oldFillOpacity(0)
@@ -112,16 +100,16 @@ void QSvgFillStyle::setFillOpacity(qreal
     m_fillOpacity = opacity;
 }

-void QSvgFillStyle::setFillStyle(QSvgPaintStyleProperty* style)
+void QSvgFillStyle::setPaintServer(QSvgPaintServerSharedPtr paintServer)
 {
-    m_style = style;
+    m_paintServer = paintServer;
     m_fillSet = 1;
 }

 void QSvgFillStyle::setBrush(QBrush brush)
 {
     m_fill = std::move(brush);
-    m_style = nullptr;
+    m_paintServer.reset();
     m_fillSet = 1;
 }

@@ -134,8 +122,8 @@ void QSvgFillStyle::apply(QPainter *p, c
     if (m_fillRuleSet)
         states.fillRule = m_fillRule;
     if (m_fillSet) {
-        if (m_style)
-            p->setBrush(m_style->brush(p, n, states));
+        if (m_paintServer)
+            p->setBrush(m_paintServer->brush(p, n, states));
         else
             p->setBrush(m_fill);
     }
@@ -247,7 +235,6 @@ QSvgStrokeStyle::QSvgStrokeStyle()
     , m_oldStrokeOpacity(0.0)
     , m_strokeDashOffset(0)
     , m_oldStrokeDashOffset(0)
-    , m_style(0)
     , m_vectorEffect(0)
     , m_oldVectorEffect(0)
     , m_strokeSet(0)
@@ -286,8 +273,8 @@ void QSvgStrokeStyle::apply(QPainter *p,
         states.vectorEffect = m_vectorEffect;

     if (m_strokeSet) {
-        if (m_style)
-            pen.setBrush(m_style->brush(p, n, states));
+        if (m_paintServer)
+            pen.setBrush(m_paintServer->brush(p, n, states));
         else
             pen.setBrush(m_stroke.brush());
     }
@@ -372,56 +359,6 @@ void QSvgStrokeStyle::setDashArray(const
     m_strokeDashArraySet = 1;
 }

-QSvgSolidColorStyle::QSvgSolidColorStyle(const QColor &color)
-    : m_solidColor(color)
-{
-}
-
-QSvgGradientStyle::QSvgGradientStyle(QGradient *grad)
-    : m_gradient(grad), m_gradientStopsSet(false)
-{
-}
-
-QBrush QSvgGradientStyle::brush(QPainter *, const QSvgNode *, QSvgExtraStates &)
-{
-    if (!m_link.isEmpty()) {
-        resolveStops();
-    }
-
-    // If the gradient is marked as empty, insert transparent black
-    if (!m_gradientStopsSet) {
-        m_gradient->setStops(QGradientStops() << QGradientStop(0.0, QColor(0, 0, 0, 0)));
-        m_gradientStopsSet = true;
-    }
-
-    QBrush b(*m_gradient);
-
-    if (!m_transform.isIdentity())
-        b.setTransform(m_transform);
-
-    return b;
-}
-
-
-void QSvgGradientStyle::setTransform(const QTransform &transform)
-{
-    m_transform = transform;
-}
-
-QSvgPatternStyle::QSvgPatternStyle(QSvgPattern *pattern)
-    : m_pattern(pattern)
-{
-
-}
-
-QBrush QSvgPatternStyle::brush(QPainter *p, const QSvgNode *node, QSvgExtraStates &states)
-{
-    m_patternImage = m_pattern->patternImage(p, states, node);
-    QBrush b(m_patternImage);
-    b.setTransform(m_pattern->appliedTransform());
-    return b;
-}
-
 QSvgTransformStyle::QSvgTransformStyle(const QTransform &trans)
     : m_transform(trans)
 {
@@ -463,21 +400,6 @@ QSvgStyleProperty::Type QSvgStrokeStyle:
     return STROKE;
 }

-QSvgStyleProperty::Type QSvgSolidColorStyle::type() const
-{
-    return SOLID_COLOR;
-}
-
-QSvgStyleProperty::Type QSvgGradientStyle::type() const
-{
-    return GRADIENT;
-}
-
-QSvgStyleProperty::Type QSvgPatternStyle::type() const
-{
-    return PATTERN;
-}
-
 QSvgStyleProperty::Type QSvgTransformStyle::type() const
 {
     return TRANSFORM;
@@ -925,36 +847,4 @@ QSvgStyleProperty::Type QSvgOpacityStyle
     return OPACITY;
 }

-void QSvgGradientStyle::setStopLink(const QString &link, QSvgTinyDocument *doc)
-{
-    m_link = link;
-    m_doc  = doc;
-}
-
-void QSvgGradientStyle::resolveStops()
-{
-    QStringList visited;
-    resolveStops_helper(&visited);
-}
-
-void QSvgGradientStyle::resolveStops_helper(QStringList *visited)
-{
-    if (!m_link.isEmpty() && m_doc) {
-        QSvgStyleProperty *prop = m_doc->styleProperty(m_link);
-        if (prop && !visited->contains(m_link)) {
-            visited->append(m_link);
-            if (prop->type() == QSvgStyleProperty::GRADIENT) {
-                QSvgGradientStyle *st =
-                    static_cast<QSvgGradientStyle*>(prop);
-                st->resolveStops_helper(visited);
-                m_gradient->setStops(st->qgradient()->stops());
-                m_gradientStopsSet = st->gradientStopsSet();
-            }
-        } else {
-            qWarning("Could not resolve property : %s", qPrintable(m_link));
-        }
-        m_link = QString();
-    }
-}
-
 QT_END_NAMESPACE
Index: qt6-svg/src/svg/qsvgtinydocument.cpp
===================================================================
--- qt6-svg.orig/src/svg/qsvgtinydocument.cpp
+++ qt6-svg/src/svg/qsvgtinydocument.cpp
@@ -378,17 +378,20 @@ QSvgNode *QSvgTinyDocument::namedNode(co
     return m_namedNodes.value(id);
 }

-void QSvgTinyDocument::addNamedStyle(const QString &id, QSvgPaintStyleProperty *style)
+void QSvgTinyDocument::addPaintServer(QSvgPaintServerSharedPtr paintServer, const QString &id)
 {
-    if (!m_namedStyles.contains(id))
-        m_namedStyles.insert(id, style);
+    if (id.isEmpty())
+        return;
+
+    if (!m_paintServers.contains(id))
+        m_paintServers.insert(id, paintServer);
     else
         qCWarning(lcSvgHandler) << "Duplicate unique style id:" << id;
 }

-QSvgPaintStyleProperty *QSvgTinyDocument::namedStyle(const QString &id) const
+QSvgPaintServerSharedPtr QSvgTinyDocument::paintServer(QStringView id) const
 {
-    return m_namedStyles.value(id);
+    return m_paintServers.value(id.toString());
 }

 void QSvgTinyDocument::restartAnimation()
Index: qt6-svg/src/svg/qsvgtinydocument_p.h
===================================================================
--- qt6-svg.orig/src/svg/qsvgtinydocument_p.h
+++ qt6-svg/src/svg/qsvgtinydocument_p.h
@@ -25,6 +25,7 @@
 #include "QtCore/qxmlstream.h"
 #include "qsvgstyle_p.h"
 #include "qsvgfont_p.h"
+#include <QtSvg/private/qsvgpaintserver_p.h>

 QT_BEGIN_NAMESPACE

@@ -76,8 +77,8 @@ public:
     QSvgFont *svgFont(const QString &family) const;
     void addNamedNode(const QString &id, QSvgNode *node);
     QSvgNode *namedNode(const QString &id) const;
-    void addNamedStyle(const QString &id, QSvgPaintStyleProperty *style);
-    QSvgPaintStyleProperty *namedStyle(const QString &id) const;
+    void addPaintServer(QSvgPaintServerSharedPtr paintServer, const QString &id);
+    QSvgPaintServerSharedPtr paintServer(QStringView id) const;

     void restartAnimation();
     inline int currentElapsed() const;
@@ -101,7 +102,7 @@ private:

     QHash<QString, QSvgRefCounter<QSvgFont> > m_fonts;
     QHash<QString, QSvgNode *> m_namedNodes;
-    QHash<QString, QSvgRefCounter<QSvgPaintStyleProperty> > m_namedStyles;
+    QHash<QString, QSvgPaintServerSharedPtr> m_paintServers;

     qint64 m_time;
     bool  m_animated;
