Index: qt6-svg/src/svg/qsvghandler.cpp
===================================================================
--- qt6-svg.orig/src/svg/qsvghandler.cpp
+++ qt6-svg/src/svg/qsvghandler.cpp
@@ -4645,8 +4645,7 @@ void QSvgHandler::parse()
             // namespaceUri is empty. The only possible strategy at
             // this point is to do what everyone else seems to do and
             // ignore the reported namespaceUri completely.
-            if (remainingUnfinishedElements
-                    && startElement(xml->name().toString(), xml->attributes())) {
+            if (remainingUnfinishedElements && startElement(xml->name().toString(), xml->attributes())) {
                 --remainingUnfinishedElements;
             } else {
                 delete m_doc;
Index: qt6-svg/src/svg/qsvgstructure.cpp
===================================================================
--- qt6-svg.orig/src/svg/qsvgstructure.cpp
+++ qt6-svg/src/svg/qsvgstructure.cpp
@@ -720,6 +720,7 @@ QSvgPattern::QSvgPattern(QSvgNode *paren
     m_rect(bounds),
     m_viewBox(viewBox),
     m_contentUnits(contentUnits),
+    m_isRendering(false),
     m_transform(transform)
 
 {
@@ -805,6 +806,13 @@ QImage QSvgPattern::renderPattern(QSize
     }
     pattern.fill(Qt::transparent);
 
+    if (m_isRendering) {
+        qCWarning(lcSvgDraw) << "The pattern is trying to render itself recursively. "
+                                "Returning a transparent QImage of the right size.";
+        return pattern;
+    }
+
+    QScopedValueRollback<bool> guard(m_isRendering, true);
     // Draw the pattern using our QPainter.
     QPainter patternPainter(&pattern);
     QSvgExtraStates patternStates;
Index: qt6-svg/src/svg/qsvgstructure_p.h
===================================================================
--- qt6-svg.orig/src/svg/qsvgstructure_p.h
+++ qt6-svg/src/svg/qsvgstructure_p.h
@@ -238,6 +238,7 @@ private:
     QSvgRectF m_rect;
     QRectF m_viewBox;
     QtSvg::UnitTypes m_contentUnits;
+    mutable bool m_isRendering;
     QTransform m_transform;
 };
 
