Author: Tian Shilin <tianshilin@uniontech.com>
Date:   Tue Mar 17 10:39:41 2026
Subject: Unify the way resource identifiers are parsed
Upstream: https://codereview.qt-project.org/c/qt/qtsvg/+/722204


---
Index: qt6-svg/src/svg/qsvghandler.cpp
===================================================================
--- qt6-svg.orig/src/svg/qsvghandler.cpp
+++ qt6-svg/src/svg/qsvghandler.cpp
@@ -774,34 +774,42 @@ static QList<qreal> parsePercentageList(
     return points;
 }

-static QString idFromUrl(const QString &url)
+/**
+ *  The form is <IRI>. This function parses local
+ *  IRI references, i.e, resources referenced within
+ *  the current document. e.g, href = "#id"
+*/
+static QString idFromIRI(QStringView iri)
 {
-    // The form is url(<IRI>), where IRI can be
-    // just an ID on #<id> form.
-    QString::const_iterator itr = url.constBegin();
-    QString::const_iterator end = url.constEnd();
-    QString id;
-    while (itr != end && (*itr).isSpace())
-        ++itr;
-    if (itr != end && (*itr) == QLatin1Char('('))
-        ++itr;
-    else
+    iri = iri.trimmed();
+
+    if (!iri.startsWith(QLatin1Char('#')))
         return QString();
-    while (itr != end && (*itr).isSpace())
-        ++itr;
-    if (itr != end && (*itr) == QLatin1Char('#')) {
-        id += *itr;
-        ++itr;
-    } else {
+
+    return iri.sliced(1).toString();
+}
+
+/**
+ *  The form is <FuncIRI>, where FuncIRI takes
+ *  the form of url(<IRI>). This syntax is used
+ *  in properties that accept both strings and
+ *  IRIs, eliminating any ambiguity. e.g, fill = "url(#id)"
+*/
+static QString idFromFuncIRI(QStringView iri)
+{
+    iri = iri.trimmed();
+
+    if (!iri.startsWith(QLatin1String("url(")))
         return QString();
-    }
-    while (itr != end && (*itr) != QLatin1Char(')')) {
-        id += *itr;
-        ++itr;
-    }
-    if (itr == end || (*itr) != QLatin1Char(')'))
+
+    iri.slice(4);
+
+    const qsizetype closingBracePos = iri.indexOf(QLatin1Char(')'));
+    if (closingBracePos == -1)
         return QString();
-    return id;
+
+    iri = iri.first(closingBracePos);
+    return idFromIRI(iri);
 }

 /**
@@ -999,9 +1007,9 @@ static void parseColor(QSvgNode *,
     }
 }

-static QSvgStyleProperty *styleFromUrl(QSvgNode *node, const QString &url)
+static QSvgStyleProperty *styleFromUrl(QSvgNode *node, QStringView url)
 {
-    return node ? node->styleProperty(idFromUrl(url)) : 0;
+    return node ? node->styleProperty(idFromFuncIRI(url)) : 0;
 }

 static void parseBrush(QSvgNode *node,
@@ -1027,14 +1035,14 @@ static void parseBrush(QSvgNode *node,
         //fill attribute handling
         if ((!attributes.fill.isEmpty()) && (attributes.fill != QT_INHERIT) ) {
             if (attributes.fill.size() > 3 && attributes.fill.mid(0, 3) == QLatin1String("url")) {
-                QString value = attributes.fill.mid(3, attributes.fill.size() - 3).toString();
+                QStringView value = attributes.fill;
                 QSvgStyleProperty *style = styleFromUrl(node, value);
                 if (style) {
                     if (style->type() == QSvgStyleProperty::SOLID_COLOR || style->type() == QSvgStyleProperty::GRADIENT
                             || style->type() == QSvgStyleProperty::PATTERN)
                         prop->setFillStyle(reinterpret_cast<QSvgPaintStyleProperty *>(style));
                 } else {
-                    QString id = idFromUrl(value);
+                    QString id = idFromFuncIRI(value);
                     prop->setPaintStyleId(id);
                     handler->pushUnresolvedStyle(prop);
                 }
@@ -1197,14 +1205,14 @@ static void parsePen(QSvgNode *node,
         //stroke attribute handling
         if ((!attributes.stroke.isEmpty()) && (attributes.stroke != QT_INHERIT) ) {
             if (attributes.stroke.size() > 3 && attributes.stroke.mid(0, 3) == QLatin1String("url")) {
-                 QString value = attributes.stroke.mid(3, attributes.stroke.size() - 3).toString();
+                    QStringView value = attributes.stroke;
                     QSvgStyleProperty *style = styleFromUrl(node, value);
                     if (style) {
                         if (style->type() == QSvgStyleProperty::SOLID_COLOR || style->type() == QSvgStyleProperty::GRADIENT
                             || style->type() == QSvgStyleProperty::PATTERN)
                         prop->setStyle(reinterpret_cast<QSvgPaintStyleProperty *>(style));
                     } else {
-                        QString id = idFromUrl(value);
+                        QString id = idFromFuncIRI(value);
                         prop->setPaintStyleId(id);
                         handler->pushUnresolvedStyle(prop);
                     }
@@ -2303,56 +2311,29 @@ static void parseExtendedAttributes(QSvg
         return;

     if (!attributes.mask.isEmpty()) {
-        QString maskStr = attributes.mask.toString().trimmed();
-        if (maskStr.size() > 3 && maskStr.mid(0, 3) == QLatin1String("url"))
-            maskStr = maskStr.mid(3, maskStr.size() - 3);
-        QString maskId = idFromUrl(maskStr);
-        if (maskId.startsWith(QLatin1Char('#'))) //TODO: handle urls and ids in a single place
-            maskId.remove(0, 1);
-
+        QString maskId = idFromFuncIRI(attributes.mask);
         node->setMaskId(maskId);
     }

     if (!attributes.markerStart.isEmpty() &&
         !handler->options().testFlag(QtSvg::Tiny12FeaturesOnly)) {
-        QString markerStr = attributes.markerStart.toString().trimmed();
-        if (markerStr.size() > 3 && markerStr.mid(0, 3) == QLatin1String("url"))
-            markerStr = markerStr.mid(3, markerStr.size() - 3);
-        QString markerId = idFromUrl(markerStr);
-        if (markerId.startsWith(QLatin1Char('#'))) //TODO: handle urls and ids in a single place
-            markerId.remove(0, 1);
+        QString markerId = idFromFuncIRI(attributes.markerStart);
         node->setMarkerStartId(markerId);
     }
     if (!attributes.markerMid.isEmpty() &&
         !handler->options().testFlag(QtSvg::Tiny12FeaturesOnly)) {
-        QString markerStr = attributes.markerMid.toString().trimmed();
-        if (markerStr.size() > 3 && markerStr.mid(0, 3) == QLatin1String("url"))
-            markerStr = markerStr.mid(3, markerStr.size() - 3);
-        QString markerId = idFromUrl(markerStr);
-        if (markerId.startsWith(QLatin1Char('#'))) //TODO: handle urls and ids in a single place
-            markerId.remove(0, 1);
+        QString markerId = idFromFuncIRI(attributes.markerMid);
         node->setMarkerMidId(markerId);
     }
     if (!attributes.markerEnd.isEmpty() &&
         !handler->options().testFlag(QtSvg::Tiny12FeaturesOnly)) {
-        QString markerStr = attributes.markerEnd.toString().trimmed();
-        if (markerStr.size() > 3 && markerStr.mid(0, 3) == QLatin1String("url"))
-            markerStr = markerStr.mid(3, markerStr.size() - 3);
-        QString markerId = idFromUrl(markerStr);
-        if (markerId.startsWith(QLatin1Char('#'))) //TODO: handle urls and ids in a single place
-            markerId.remove(0, 1);
+        QString markerId = idFromFuncIRI(attributes.markerEnd);
         node->setMarkerEndId(markerId);
     }

     if (!attributes.filter.isEmpty() &&
         !handler->options().testFlag(QtSvg::Tiny12FeaturesOnly)) {
-        QString filterStr = attributes.filter.toString().trimmed();
-
-        if (filterStr.size() > 3 && filterStr.mid(0, 3) == QLatin1String("url"))
-            filterStr = filterStr.mid(3, filterStr.size() - 3);
-        QString filterId = idFromUrl(filterStr);
-        if (filterId.startsWith(QLatin1Char('#'))) //TODO: handle urls and ids in a single place
-            filterId.remove(0, 1);
+        QString filterId = idFromFuncIRI(attributes.filter);
         node->setFilterId(filterId);
     }

@@ -2950,7 +2931,7 @@ static void parseBaseGradient(QSvgNode *
                               QSvgGradientStyle *gradProp,
                               QSvgHandler *handler)
 {
-    QString link   = attributes.value(QLatin1String("xlink:href")).toString();
+    QString linkId = idFromIRI(attributes.value(QLatin1String("xlink:href")));
     QStringView trans  = attributes.value(QLatin1String("gradientTransform"));
     QString spread = attributes.value(QLatin1String("spreadMethod")).toString();
     QString units = attributes.value(QLatin1String("gradientUnits")).toString();
@@ -2965,9 +2946,9 @@ static void parseBaseGradient(QSvgNode *

     QTransform matrix;
     QGradient *grad = gradProp->qgradient();
-    if (node && !link.isEmpty()) {
-        QSvgStyleProperty *prop = node->styleProperty(link);
-        //qDebug()<<"inherited "<<prop<<" ("<<link<<")";
+    if (node && !linkId.isEmpty()) {
+        QSvgStyleProperty *prop = node->styleProperty(linkId);
+        //qDebug()<<"inherited "<<prop<<" ("<<linkId<<")";
         if (prop && prop->type() == QSvgStyleProperty::GRADIENT) {
             QSvgGradientStyle *inherited =
                 static_cast<QSvgGradientStyle*>(prop);
@@ -2980,7 +2961,7 @@ static void parseBaseGradient(QSvgNode *

             matrix = inherited->qtransform();
         } else {
-            gradProp->setStopLink(link, handler->document());
+            gradProp->setStopLink(linkId, handler->document());
         }
     }

@@ -4221,13 +4202,13 @@ static QSvgNode *createUseNode(QSvgNode
                                const QXmlStreamAttributes &attributes,
                                QSvgHandler *handler)
 {
-    QString linkId = attributes.value(QLatin1String("xlink:href")).toString().remove(0, 1);
+    QString linkId = idFromIRI(attributes.value(QLatin1String("xlink:href")));
     const QStringView xStr = attributes.value(QLatin1String("x"));
     const QStringView yStr = attributes.value(QLatin1String("y"));
     QSvgStructureNode *group = nullptr;

     if (linkId.isEmpty())
-        linkId = attributes.value(QLatin1String("href")).toString().remove(0, 1);
+        linkId = idFromIRI(attributes.value(QLatin1String("href")));
     switch (parent->type()) {
     case QSvgNode::Doc:
     case QSvgNode::Defs:
Index: qt6-svg/tests/auto/qsvgrenderer/tst_qsvgrenderer.cpp
===================================================================
--- qt6-svg.orig/tests/auto/qsvgrenderer/tst_qsvgrenderer.cpp
+++ qt6-svg/tests/auto/qsvgrenderer/tst_qsvgrenderer.cpp
@@ -135,6 +135,9 @@ void tst_QSvgRenderer::getSetCheck()
 void tst_QSvgRenderer::inexistentUrl()
 {
     const char *src = "<svg><g><path d=\"\" style=\"stroke:url(#inexistent)\"/></g></svg>";
+#if QT_CONFIG(regularexpression)
+    QTest::ignoreMessage(QtWarningMsg, QRegularExpression("Could not resolve property: inexistent"));
+#endif

     QByteArray data(src);
     QSvgRenderer renderer(data);
@@ -2021,7 +2024,7 @@ void tst_QSvgRenderer::testMisplacedElem

     QTest::ignoreMessage(QtWarningMsg, "<input>:2:68: Could not add child element to parent "
                                        "element because the types are incorrect.");
-    QTest::ignoreMessage(QtWarningMsg, "<input>:4:28: Could not resolve property: #ptn");
+    QTest::ignoreMessage(QtWarningMsg, "<input>:4:28: Could not resolve property: ptn");

     QSvgRenderer renderer(svg);
     QPainter painter(&image);
Index: qt6-svg/src/svg/qsvgnode.cpp
===================================================================
--- qt6-svg.orig/src/svg/qsvgnode.cpp
+++ qt6-svg/src/svg/qsvgnode.cpp
@@ -339,11 +339,8 @@ QSvgStyleProperty * QSvgNode::stylePrope

 QSvgPaintStyleProperty * QSvgNode::styleProperty(const QString &id) const
 {
-    QString rid = id;
-    if (rid.startsWith(QLatin1Char('#')))
-        rid.remove(0, 1);
     QSvgTinyDocument *doc = document();
-    return doc ? doc->namedStyle(rid) : 0;
+    return doc ? doc->namedStyle(id) : 0;
 }

 QRectF QSvgNode::internalFastBounds(QPainter *p, QSvgExtraStates &states) const
